site/ja/guide/create_op.md

```diff
@@ -0,0 +1,1231 @@
+```
+
+> 前述したように、もしgcc>=5でコンパイルする場合は、bazelのコマンドラインに `--cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0"` を追加してください。
+
+> Note: 標準の `cc_library` ルールを利用して共有ライブラリ（`.so` ファイル）を作成できますが、`tf_custom_op_library` マクロを利用することを強く推奨します。このマクロは、必要となるいくつかの依存関係を追加し、共有ライブラリがTensorFlowのプラグイン読み込み機構と適合しているかを確認します。
+
+
+## オペレーションをPythonで使用する
+
+TensorFlowのPython APIは、動的ライブラリをロードしてTensorFlowのフレームワークにオペレーションを登録する、`tf.load_op_library` 関数を提供しています。
+`load_op_library` は、オペレーションとカーネルのPythonラッパーを含んだ、Pythonモジュールを返します。
+たとえば、オペレーションを1度ビルドしたら、Pythonから次のように実行できます。
+
+```python
+import tensorflow as tf
+zero_out_module = tf.load_op_library('./zero_out.so')
+with tf.Session(''):
+  zero_out_module.zero_out([[1, 2], [3, 4]]).eval()
+
+# Prints
+array([[1, 0], [0, 0]], dtype=int32)
+```
+
+([PEP8](https://www.python.org/dev/peps/pep-0008/) に従うために、)生成された関数は、スネークケースの名前が与えられることを覚えておいてください。
+つまり、C++のファイル内で `ZeroOut` と名付けられたオペレーションは、Pythonの関数では `zero_out` で呼ぶことになるでしょう。
+
+Pythonモジュールから `import` 可能な、規則正しい関数としてオペレーションを利用可能にするには、次のようにしてPythonのソースコードのファイルに、`load_op_library` の呼び出しをもつのが便利かもしれません。
+
+```python
+import tensorflow as tf
+
+zero_out_module = tf.load_op_library('./zero_out.so')
+zero_out = zero_out_module.zero_out
+```
+
+
+## オペレーションが動作を検証する
+
+正しくオペレーションが実装できたことを確かめるよい方法は、テストを書くことです。
+次の内容を持った `zero_out_op_test.py` を作ってください。
+
+```python
+import tensorflow as tf
+
+class ZeroOutTest(tf.test.TestCase):
+  def testZeroOut(self):
+    zero_out_module = tf.load_op_library('./zero_out.so')
+    with self.test_session():
+      result = zero_out_module.zero_out([5, 4, 3, 2, 1])
+      self.assertAllEqual(result.eval(), [5, 0, 0, 0, 0])
+
+if __name__ == "__main__":
+  tf.test.main()
+```
+
+そして、テストを実行してください（TensorFlowがインストール済みであると想定しています）。
+
+
+## より発展的な機能をオペレーションに構築する
+
+基本的な（そして、やや制限された）オペレーションと実装の構築方法を学んだため、オペレーションに対して構築する必要が出てくるであろう、一般的でより複雑なことを見ていきましょう。
+
+* [条件のチェックと検証](#条件チェックと検証)
+* [オペレーションの登録](#オペレーションの登録)
+  * [アトリビュート](#アトリビュート)
+  * [アトリビュートタイプ](#アトリビュート型)
+  * [ポリモーフィズム](#ポリモーフィズム)
+  * [入力と出力](#入力と出力)
+  * [後方互換性](#後方互換性)
+* [GPUサポート](#GPUサポート)
+  * [GPUデバイス向けのカーネルコンパイル](#GPUデバイス向けのカーネルコンパイル)
+* [Pythonにおける勾配の実装](#Pythonにおける勾配の実装)
+* [C++でのシェイプ推論](#C++でのシェイプ推論)
+
+
+### 条件チェックと検証
+
+これまで想定してきた例では、いかなるシェイプのテンソルに対しても適用できるオペレーションを想定していました。
+仮にベクトルに対してのみ適用したい場合、どうなるでしょうか？
+これまでのOpKernelの実装に、チェックを追加することになります。
+
+```c++
+  void Compute(OpKernelContext* context) override {
+    // 入力テンソルを取得する
+    const Tensor& input_tensor = context->input(0);
+
+    OP_REQUIRES(context, TensorShapeUtils::IsVector(input_tensor.shape()),
+                errors::InvalidArgument("ZeroOut expects a 1-D vector."));
+    // ...
+  }
+```
+
+これは入力がベクトルであることを主張し、もしそうでないなら `InvalidArgument` ステータスを設定して戻ります。
+[`OP_REQUIRES` マクロ][validation-macro] は、3つの引数を受け取ります。
+
+* `SetStatus()` メソッドのための、`OpKernelContext` または `OpKernelConstruction` のポインタになる `context` ([`tensorflow/core/framework/op_kernel.h`](https://www.tensorflow.org/code/tensorflow/core/framework/op_kernel.h) を確認してください)。
```

Comment:
[weak suggestion]
```suggestion
*  `context`、`SetStatus()` メソッドのための、`OpKernelContext` または `OpKernelConstruction` のポインタです ([`tensorflow/core/framework/op_kernel.h`](https://www.tensorflow.org/code/tensorflow/core/framework/op_kernel.h) を確認してください)。
```

Reply:
Thanks!
I think the original sentence **`context`,** should be **`context`.** (not comma)